src/components/scripts/oauth.mjs

```diff
@@ -0,0 +1,221 @@
+    }
+
+    // Add form to page and submit it to open the OAuth 2.0 endpoint.
+    document.body.appendChild(form);
+    form.submit();
+  }
+
+  async #runHttpServer() {
+    const server =
+      await this.#socketServer.opened;
+    const connections = server.readable.getReader();
+
+    while (true) {
+      const { value: connection, done } = await connections.read();
+
+      // Send the connection to the callback
+      if (connection) {
+        this.#connectionReceived(connection);
+      }
+
+      // Release the connection if we're done
+      if (done) {
+        connections.releaseLock();
+        break;
+      }
+    }
+
+    // Wait for the server to be closed
+    await this.#socketServer.closed;
+    this.#socketServer = undefined;
+  }
+
+  async #connectionReceived(connection) {
+    return connection.opened.then(async (socket) => {
+      const reader = socket.readable.getReader();
+      const value = await this.#readStream(reader);
+      const { success, html } = this.#generateResponse(value);
+      await this.#writeStream(socket, html);
+
+      connection.close();
+
+      if (success) {
+        const accessCode = await this.#getAccessCode();
+        this.#promiseResolver(accessCode);
+      }
+    });
+  }
+
+  #generateResponse(value) {
+    const text = new TextDecoder().decode(value);
+
+    if (this.#processRequest(text)) {
+      return {
+        success: true,
+        html: "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\n\r\n<h1>You can close this page</h1>",
+      };
+    } else {
+      return { success: false, html: "HTTP/1.1 400 KO" };
+    }
+  }
+
+  #processRequest(text) {
+    const lines = text.split("\r\n");
+
+    if (lines[0].includes("state=pass-through%20value")) {
+      let matches = lines[0].match(/code=(.*)[&$]/);
+      this.#authenticationCode = matches[1];
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  async #getAccessCode() {
+    // See https://developers.google.com/identity/protocols/oauth2/native-app#exchange-authorization-code
+
+    const url_params =
+      `?code=${this.#authenticationCode}&` +
+      `client_id=${this.#clientId}&` +
+      `client_secret=${this.#clientSecret}&` +
+      `redirect_uri=${this.#redirectUri}&` +
+      `grant_type=authorization_code`;
+
+    const url = this.#token_endpoint + url_params;
+
+    return fetch(url, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded",
+      },
+    })
+      .then((response) => {
+        if (response.status >= 200 && response.status < 300) {
+          return Promise.resolve(response.json());
+        } else {
+          return Promise.reject(
+            new Error(response.status.toString + " - " + response.statusText),
+          );
+        }
+      })
+      .then((json) => {
+        this.#accessCode = json.access_token;
+        this.#refreshCode = json.refresh_token;
+
+        return this.#accessCode;
+      });
```

Comment:
```suggestion
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
    });

    if (response.status >= 200 && response.status < 300) {
      const json = await response.json();
      this.#accessCode = json.access_token;
      this.#refreshCode = json.refresh_token;
      return this.#accessCode;
    }

    throw new Error(response.status.toString + ' - ' + response.statusText);
```

Because we're in an `async` function here, we can rewrite the `.then` chaining of the fetch into `await`s to make it what's happening easier to manage